### PR TITLE
[SPMD] index local shards correctly in replication

### PR DESCRIPTION
**Pull request by yeounoh**
Replication across multi-host devices was buggy due to incorrect index mapping. Tested locally with 2x v4-8 slice.

### CODE REVIEW

File: third_party/xla_client/pjrt_computation_client.cc [hunk 1]
```diff
@@ -315,17 +315,25 @@ ComputationClient::DataPtr PjRtComputationClient::ReplicateShardedData(
     XLA_CHECK_EQ(shards.size(), GetLocalDevices().size());
     std::vector<std::vector<ComputationClient::DataPtr>> arguments_by_device(
         GetLocalDevices().size(), std::vector<ComputationClient::DataPtr>(1));
-    for (auto shard : shards) {
+    for (int i = 0; i < shards.size(); ++i) {
```
Comment (jonb377) [inline third_party/xla_client/pjrt_computation_client.cc:318]: Can we reuse ShardingUtil::InputHandler here?

Comment (yeounoh) [reply third_party/xla_client/pjrt_computation_client.cc:318]: +1 let's do that

Comment (yeounoh) [reply third_party/xla_client/pjrt_computation_client.cc]: ah, forgot about the circular dependencies issue. We should try avoid using the xla sharding util in the runtime client.

File: third_party/xla_client/pjrt_computation_client.cc [hunk 2]
```diff
@@ -315,17 +315,25 @@ ComputationClient::DataPtr PjRtComputationClient::ReplicateShardedData(
...
+      auto shard = shards[i];
       std::vector<std::string> device_spec =
           absl::StrSplit(shard->device(), ':');
       XLA_CHECK_EQ(device_spec.size(), 2)
           << "Invalid device specification: " << shard->device();
-      int device_i = std::stoi(device_spec[1]);
-      arguments_by_device[device_i][0] = shard;
+      TF_VLOG(3) << shard->device() << " is mapped to local device index " << i;
+      arguments_by_device[i][0] = shard;
```
Comment (jonb377) [inline third_party/xla_client/pjrt_computation_client.cc:325]: Using `i` directly here requires that the order of the shards within PjRtShardedData corresponds to the order of devices returned from `GetLocalDevices` - is that a safe assumption? In the `InputHandler`, we explicitly [remap](https://github.com/pytorch/xla/blob/09d6775d4c42ce1d0582ef5e797e8bd741166ffb/torch_xla/csrc/xla_sharding_util.cpp#L226) the shard index to the device index.

In my `load_local_shards` [PR](https://github.com/pytorch/xla/pull/5047/files#diff-12cff199d1701b121499c962ef4dabbe2bba7b134cae3f6b54d8082a52a1c85aR1659), the order within PjRtShardedData will depend on the order specified by the user, but other than that I don't think we ever change the order - we just don't have any checks that the order matches.

**Review by jonb377 — APPROVED** [review]
LGTM, thanks Yeounoh!

File: third_party/xla_client/pjrt_computation_client.cc (hunk 2 above)
Comment (yeounoh) [reply third_party/xla_client/pjrt_computation_client.cc]: Yea, we should try to explicitly remap.

**Pull request merged**